sys-process/bottom: New package

bowlofeggs opened the pull request:
Package-Manager: Portage-3.0.20, Repoman-3.0.2
Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+# Copyright 2020-2021 Gentoo Authors
```
did that ebuild exist in 2020? if not, should be just 20201

bowlofeggs replied:
This ebuild (or rather, an ancestor of this file) is currently part of `::guru`, and @juippis had written to suggest moving it to `::gentoo`. It was originally written in 2020:

https://cgit.gentoo.org/repo/proj/guru.git/commit/sys-process/bottom?id=615d58ffcdd2233e4e5a3baaafea35d32afa5d61

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+# Copyright 2020-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# Auto-Generated by cargo-ebuild 0.3.2
```
remove that line.

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+SRC_URI="$(cargo_crate_uris ${CRATES})"
+
+# License set may be more restrictive as OR is not respected
+# use cargo-license for a more accurate license picture
```
clean out those lines

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+
+# Auto-Generated by cargo-ebuild 0.3.2
+
+EAPI=7
```
you can make it `EAPI=8`
it will force rust-1.53 dependency though, which is not stable yet. but it will go stable soon.

gyakovlev reviewed (changes requested):
looks quite good, left some nitty comments.

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+SLOT="0"
+KEYWORDS="~amd64"
+
+DOCS=( CHANGELOG.md README.md )
```
move DOCS to `src_install` and make it `local`, just before calling `einstalldocs`.

gyakovlev commented on sys-process/bottom/bottom-0.6.3.ebuild:
```diff
@@ -0,0 +1,199 @@
+
+DESCRIPTION="A graphical process/system monitor with a customizable interface"
+HOMEPAGE="https://github.com/ClementTsang/bottom"
+SRC_URI="$(cargo_crate_uris ${CRATES})"
```
you can just leave it as `SRC_URI="$(cargo_crate_uris)"`

it will pick up `CRATES` automatically since some time ago.

gyakovlev commented:
also looks like this package produces shell completions during build, maybe install those?
look here how they find it (note for gentoo it MAY be in target/debug and target/release, depending if `USE=debug` is set)
https://github.com/ClementTsang/bottom/blob/master/deployment/linux/arch/PKGBUILD.template
I also find some files in ripgrep ebuild: https://github.com/gentoo/gentoo/blob/7dfadba6743d39de86ab0fb043a334d2e1f4d893/sys-apps/ripgrep/ripgrep-13.0.0.ebuild#L102
you can use ripgrep approach, but check for `btm.bash` and not `ripgrep-stamp`

bowlofeggs commented:
Hello @gyakovlev, and thank you for the review. I am going to rebase my commits, apply the following patch to address (most of) your comments, and force push:

```
diff --git a/sys-process/bottom/bottom-0.6.3.ebuild b/sys-process/bottom/bottom-0.6.3.ebuild
index 3a951ae8994..d2971cb3738 100644
--- a/sys-process/bottom/bottom-0.6.3.ebuild
+++ b/sys-process/bottom/bottom-0.6.3.ebuild
@@ -1,8 +1,6 @@
 # Copyright 2020-2021 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-# Auto-Generated by cargo-ebuild 0.3.2
-
 EAPI=7
 
 CRATES="
@@ -175,25 +173,34 @@ CRATES="
        winapi-x86_64-pc-windows-gnu-0.4.0
 "
 
-inherit cargo
+inherit bash-completion-r1 cargo
 
 DESCRIPTION="A graphical process/system monitor with a customizable interface"
 HOMEPAGE="https://github.com/ClementTsang/bottom"
-SRC_URI="$(cargo_crate_uris ${CRATES})"
+SRC_URI="$(cargo_crate_uris)"
 
-# License set may be more restrictive as OR is not respected
-# use cargo-license for a more accurate license picture
 LICENSE="Apache-2.0 Apache-2.0-with-LLVM-exceptions BSD BSD-2 MIT MPL-2.0 Unlicense ZLIB"
 SLOT="0"
 KEYWORDS="~amd64"
 
-DOCS=( CHANGELOG.md README.md )
-
 # Rust packages ignore CFLAGS and LDFLAGS so let's silence the QA warnings
 QA_FLAGS_IGNORED="usr/bin/btm"
 
 src_install() {
        cargo_src_install
 
+       # Find generated shell completion files. btm.bash can be present in multiple dirs if we build
+       # additional features, so grab the first match only.
+       local BUILD_DIR="$(dirname $(find target/release -name btm.bash -print -quit))"
+
+       newbashcomp "${BUILD_DIR}"/btm.bash btm
+
+       insinto /usr/share/fish/vendor_completions.d
+       doins "${BUILD_DIR}"/btm.fish
+
+       insinto /usr/share/zsh/site-functions
+       doins "${BUILD_DIR}"/_btm
+
+       local DOCS=( CHANGELOG.md README.md )
        einstalldocs
 }
```

It does not yet switch to EAPI 8 because it's late and time for sleeping. I will look at that soon.

Thanks!

bowlofeggs commented:
@gyakovlev I remembered that I could test with rust-bin and not have to wait for Rust 1.53 to compile, so I went ahead and switched to EAPI 8 as well. I also noticed that my first patch missed your note about how the debug build might install the bash completions in the debug target, so I adjusted the `find` statement for that as well. I will apply this patch and force push:

```
diff --git a/sys-process/bottom/bottom-0.6.3.ebuild b/sys-process/bottom/bottom-0.6.3.ebuild
index d2971cb3738..5a268f8904e 100644
--- a/sys-process/bottom/bottom-0.6.3.ebuild
+++ b/sys-process/bottom/bottom-0.6.3.ebuild
@@ -1,7 +1,7 @@
 # Copyright 2020-2021 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=7
+EAPI=8
 
 CRATES="
        addr2line-0.15.1
@@ -191,7 +191,7 @@ src_install() {
 
        # Find generated shell completion files. btm.bash can be present in multiple dirs if we build
        # additional features, so grab the first match only.
-       local BUILD_DIR="$(dirname $(find target/release -name btm.bash -print -quit))"
+       local BUILD_DIR="$(dirname $(find target -name btm.bash -print -quit))"
 
        newbashcomp "${BUILD_DIR}"/btm.bash btm
```

gyakovlev reviewed (approved):
lgtm!

gyakovlev commented:
i also pushed a small fix here https://github.com/gentoo/gentoo/commit/9bf946cc09a069a9a8453b6a2c8c6b86dbd41433
basically default features are for debugging only. so I disabled those for release build. saves time a bit and reduces binary size.

gyakovlev commented:
also keyworded for ~ppc64 and ~arm64

works great on ppc64, thanks!
![bottom](https://user-images.githubusercontent.com/168902/126858821-f73b31f3-a9ef-43a5-9864-8e34fb38c0ed.png)

bowlofeggs commented:
Thanks for the review @gyakovlev, I learned some things I can apply to a few of my other Rust ebuilds.
